File: variability/extrinsic/match_to_gerlumph/match_to_gerlumph.py
import os
import re
import sys
import json
import math
import sqlite3
import urllib.parse

'''
Reuiqres:
- multiple_images.json
'''

dbfile     = sys.argv[1]
map_path   = sys.argv[2]
out_path   = sys.argv[3]

f          = open(out_path+"output/multiple_images.json",'r')
input_str  = f.read()
input_str  = re.sub(re.compile("/\*.*?\*/",re.DOTALL),"",input_str)
input_str  = re.sub(re.compile("//.*?\n" ),"",input_str)
images     = json.loads(input_str)

outfile = out_path+"output/gerlumph_maps.json"



conn = sqlite3.connect(dbfile)
cur = conn.cursor()

gerlumph_maps = []
kg_sep = 0.05
for image in images:
    
    k0 = image["k"]
    g0 = image["g"]
    s0 = image["s"]
    cur.execute("SELECT id,k,g,s,(k-?)*(k-?)+(g-?)*(g-?) AS d,ABS(s-?) as ds FROM gerlumph GROUP BY d,ds HAVING d < ?*? ORDER BY d ASC,ds ASC LIMIT 10;",[k0,k0,g0,g0,s0,kg_sep,kg_sep])
    cases = cur.fetchall()

    mymap = {}
    if len(cases) == 0:
        mymap["id"]  = "none"
        mymap["k"]   = 0
        mymap["g"]   = 0
        mymap["s"]   = 0
        mymap["dkg"] = 0
        mymap["ds"]  = 0
    else: 
        mymap["id"]  = cases[0][0]
        mymap["k"]   = float(cases[0][1])
        mymap["g"]   = float(cases[0][2])
        mymap["s"]   = float(cases[0][3])
        mymap["dkg"] = math.sqrt(cases[0][4])
        mymap["ds"]  = cases[0][5]
    gerlumph_maps.append(mymap)
    
conn.close()


with open(outfile,"w") as file:
    json.dump(gerlumph_maps,file,indent=4)

missing = []
for mymap in gerlumph_maps:
    if mymap["id"] != "none":
        if not os.path.isdir(map_path+str(mymap["id"])):
            missing.append(str(mymap["id"]))
            
if len(missing) > 0:
    missing = list(dict.fromkeys(missing))
    missing = list( map(lambda x: "ids[]="+x,missing) )
    print("ATTENTION: missing GERLUMPH maps from '"+map_path+"' !",file=sys.stderr)
    print("Click the following link to download them:",file=sys.stderr)
    print("\n",file=sys.stderr)
    myurl = "gerlumph.swin.edu.au/inc/generic/put_to_cart.php?"+"&".join(missing)
    print("      http://"+urllib.parse.quote(myurl),file=sys.stderr)
    print("\n",file=sys.stderr)

File: variability/extrinsic/match_to_gerlumph/src/match_to_gerlumph.cpp
#include <fstream>
#include <sqlite3.h> 
#include <string>
#include <cmath>
#include "json/json.h"

#include "sql_callback.hpp"

int main(int argc, char* argv[]) {

  std::string dbfile = argv[1];
  std::string out_path = argv[2];
  
  // Read the multiple images' parameters from JSON
  Json::Value images;
  std::ifstream fin(out_path+"output/multiple_images.json",std::ifstream::in);
  fin >> images;
  fin.close();


  
  sqlite3* db;
  char* zErrMsg = 0;
  int rc;
  rc = sqlite3_open(dbfile.c_str(),&db);
  if( rc ) {
    fprintf(stderr,"Can't open database: %s\n",sqlite3_errmsg(db));
    return 0;
  }

  std::vector<dbEntry> entries(images.size());
  for(int q=0;q<images.size();q++){
    // Create SQL statement
    double k0 = images[q]["k"].asDouble();
    double g0 = images[q]["g"].asDouble();
    double s0 = images[q]["s"].asDouble();
    char buffer[250];
    double kg_sep = 0.05;
    sprintf(buffer,"SELECT id,k,g,s,(k-%f)*(k-%f)+(g-%f)*(g-%f) AS dkg,ABS(s-%f) AS ds,ss,res FROM gerlumph GROUP BY dkg,ds HAVING dkg < %f ORDER BY dkg ASC,ds ASC LIMIT 1;",k0,k0,g0,g0,s0,kg_sep*kg_sep);
    const char* sql = buffer;
    
    // Execute SQL statement
    dbEntry entry;
    const char* data = "Callback function called";
    rc = sqlite3_exec(db,sql,callback,&entry,&zErrMsg);
    if( rc != SQLITE_OK ) {
      fprintf(stderr,"SQL error: %s\n",zErrMsg);
      sqlite3_free(zErrMsg);
    }
    entries[q] = entry;
  }
  sqlite3_close(db);

  
  // Write output JSON
  Json::Value json_db_entries;
  for(int i=0;i<entries.size();i++){
    Json::Value entry;
    entry["id"]   = entries[i].id;
    entry["k"]    = entries[i].k;
    entry["g"]    = entries[i].g;
    entry["s"]    = entries[i].s;
    entry["dkg"]  = entries[i].dkg;
    entry["ds"]   = entries[i].ds;
    entry["width"] = entries[i].ss;
    entry["resolution"] = entries[i].res;
    json_db_entries.append(entry);
  }

  // JSON object needs to written like below because of double precision, e.g. 1.5 is written as 1.499999999999 otherwise
  std::ofstream file_maps(out_path+"output/gerlumph_maps.json");
  Json::StreamWriterBuilder wbuilder;
  wbuilder.settings_["precision"] = 6;
  std::unique_ptr<Json::StreamWriter> writer(wbuilder.newStreamWriter());
  writer->write(json_db_entries,&file_maps);
  file_maps.close();  
}
